### ISSUE TITLE
Current Theme Not Correctly Selected on Settings Page

### ISSUE THREAD

**Issue by allenlooplee** [microsoft/WindowsTemplateStudio#4118]
**Current Theme Not Correctly Selected on Settings Page**
Each time I open the Settings page, "Default" is selected in the "Choose Theme" section, even if I hard code to return AppTheme.Light. I'm currently using WTS to generate project for WPF on Windows 10 20H2. Not sure about UWP or WinUI.

**Expected behavior**
Theme selected should reflect the currently theme correctly.

Applies to the following platforms:
| UWP              | WPF              | WinUI           |
| :--------------- | :--------------- | :---------------|
| N/A | Yes |N/A |

Wizard version: v3.9.20351.2
Template version: v3.9.20351.2

**Findings**
Pay attention to the below code in SettingsViewModel. Even if I hard coded GetCurrentTheme method to return AppTheme.Light, Theme was set to AppTheme.Default after the second line is executed. I notice that this is because the RadioButton on the Settings page is bound to Theme with the default Mode (TwoWay) which causes that RadioButton to reset Theme back to AppTheme.Default each time after the second line is executed.

        public void OnNavigatedTo(object parameter)
        {
            VersionDescription = $"App2 - {_applicationInfoService.GetVersion()}";
            Theme = _themeSelectorService.GetCurrentTheme();
        }

The resolution is to explicitly set the Mode to OneWay for all three RadioButtons. In this case, the RadioButtons get their state from Theme, while you change Theme via InvokeCommandAction behavior.

Kindly fix the code accordingly.

Regards/Allen

**Comment by mrlacey** [comment]
I'm investigating

**Comment by mrlacey** [comment]
@allenlooplee I have been unable to recreate the situation you describe above.

I've tried the following:

- Create a new WPF app with WinTS, using  `NavigationView`, `MVVMToolkit`, and adding a single settings page to the Main page that is included by default. (I've also tried everything below with the `CodeBehind` "design pattern" option, and appropriate variations in the code, but still see the same behavior.)
- Run the generated app
  - See that the default theme is used
  - Open settings page and see that the default theme is selected
  - Change the theme (light or dark) and see that theme is applied.
  - Close the settings page (no theme change)
  - Reopen the settings page and see the previously chosen theme is selected
- Close the app
- Relaunch the app
  - See the previously selected theme is applied
  - Open the settings page
  - See the previously chosen theme is selected
 
The above is all as expected.

The `TwoWay` bindings do not currently trigger the saving of the selected theme option.  
You're right though, the code will function exactly the same way if a `OneWay` binding is used in the RadioButtons on the settings page.
If this line  
`IsChecked="{Binding Theme, Converter={StaticResource EnumToBooleanConverter}, ConverterParameter=Dark}">`
is changed to   
`IsChecked="{Binding Theme, Mode=OneWay, Converter={StaticResource EnumToBooleanConverter}, ConverterParameter=Dark}">`  

The above steps can be followed and the exact same behavior is seen.

However, the ViewModel will become out of sync with the View as changes to the View won't be reflected back to the ViewModel. If additional code was added to the VM that required knowledge of the currently selected Theme this could not be obtained without querying the View directly. This isn't ideal but doesn't explain the current code complexity.

Unless I'm missing something, the exact same functionality of the app/page/VM can be maintained by using the property setter to trigger the saving of the selected theme.
This is a much simpler and more natural use of the MVVM pattern as it does not unnecessarily mix events with bindings.  
By removing the Interaction Triggers from the view and modifying the VM to remove the `SetThemeCommand`, remove the `OnSetTheme` method, and modify the `Theme` property to be like this:
```cs
    public AppTheme Theme
    {
        get { return _theme; }
        set
        {
            if (SetProperty(ref _theme, value))
            {
                _themeSelectorService.SetTheme(value);
            }
        }
    }
```
the app will continue to function as above (correctly) and have the added benefits of smaller, simpler XAML and C# code, and have no dependency on the behaviors library. I do not know why the generated code doesn't already work this way and uses the more complicated approach of mapping events to commands.

_aside: @mvegaca as author of these templates, can you explain why they use InteractionTriggers rather than the property setters for determining when to save changes?_

Regarding hardcoding `GetCurrentTheme` to always return "Light". 
I've tried the following:

- Run the generated app and set the theme to dark.
- Close the app.
- Change the following method in `ThemeSelectorService.cs`
```cs
    public AppTheme GetCurrentTheme()
    {
        ////if (App.Current.Properties.Contains("Theme"))
        ////{
        ////    var themeName = App.Current.Properties["Theme"].ToString();
        ////    Enum.TryParse(themeName, out AppTheme theme);
        ////    return theme;
        ////}

        ////return AppTheme.Default;
        return AppTheme.Light;
    }
```
- Run the app again
  - See that light theme is applied
  - Open settings and see that Light theme is selected.
  - Change the theme in the settings page and see that the change is applied
  - Restart the app and see it has reverted to the light theme.

This does not produce the problems reported above.

@allenlooplee  please can you provide either: 
- a copy of a project that has the problems you're seeing;
- or detailed instructions for creating the behavior you are seeing.

Without one or other of these there's nothing more I can do.

**Comment by allenlooplee** [comment]
@mrlacey

Thanks for your detailed investigation. Here're more details.

First, I've created a new WPF project with WinTS. Nothing was modified.

[App4-new.zip](https://github.com/microsoft/WindowsTemplateStudio/files/6090850/App4-new.zip)

Here're the manual steps to reproduce the issue:
1. Run the app.
2. Go to Settings page, and notice the Default RadioButton is selected.
3. Change the theme to Light by selecting the Light RadioButton.
4. Close the app.
5. Re-run the app, and notice the app is now applied the Light theme.
6. Go to Settings page, and notice the Default RadioButton is still selected.

In the above case, I expect the Light RadioButton was selected. If you try to select Dark RadioButton in the above third step, you'll notice the app is applied Dark theme but the Dark RadioButton is not selected on the Settings page when you re-run the app.

Now let's focus on the OnNavigatedTo method of SettingsViewModel, and write the value of Theme property before and after the assignment as below:

        public void OnNavigatedTo(object parameter)
        {
            VersionDescription = $"App4 - {_applicationInfoService.GetVersion()}";
            System.Diagnostics.Debug.WriteLine($"Theme value before assignment: {Theme}");
            Theme = _themeSelectorService.GetCurrentTheme();
            System.Diagnostics.Debug.WriteLine($"Theme value after assignment: {Theme}");
        }

Perform the above step 5 and 6, and notice the below output in Output window:

        Theme value before assignment: Default
        Theme value after assignment: Default

This means even if the app is currently applied Light theme, the value of Theme property is still Default. And after I selected Dark theme, the value of it is still Default, which is not expected.

I tracked inside EnumToBooleanConverter and noticed that the wrong value was fed back to Theme property somehow. So I tried to add Mode=OneWay in the binding expression of IsCheck of Default RadioButton, and it's not stuck with the Default RadioButton.

But now comes another issue. If you select either Light or Dark RadioButton, and then select Default RadioButton, you'll notice that both Light and Default or both Dark and Default RadioButtons are selected as the below screenshot shown:

![image](https://user-images.githubusercontent.com/132692/110127446-9fbd1180-7e00-11eb-96f9-b8544405cd79.png)

This issue can be fixed by adding Mode=OneWay to all three RadioButtons.

**Comment by mrlacey** [comment]
Ok, I have a repro.
This effects MVVMLight only.
more details to follow.

**Comment by mrlacey** [comment]
So, there's a lot going on here.

I'm torn.
On the one hand, @allenlooplee if you're happy to change the binding mode and work with that then it will unblock you, remove the incorrect behavior you're seeing and allow you to continue with no major consequences.
On the other hand, I have concerns about how Settings page is configured when using MVVMLight in a WPF app. (Note to self to review review the use of SettingsViewModels in UWP and WinUI too.) **BUT,** MVVMLight support is being deprecated and so I don't know it it worth investing time discussing changes. (Maybe a review of the impact elsewhere will help make a decision.)


The cause of the underlying issue is that the VM instance is created and set as the DataContext of the View without initializing it properly and allowing all properties to have their default values. This causes the binding to indicate that the Default theme has been selected when actually this is the default value of the property. Then, once page navigation has completed, the value of the selected theme is loaded and the UI is updated. This causes both the UI and the navigation triggered code to start both competing to set and unset the selected theme and leads to the broken behavior that is seen.

There are multiple potential fixes here but I'm not sure what's best without understanding why the current code is as it is. Is it an error or a side-effect of a bigger principle I'm not aware of?

My preference is to move the initialization code from the `OnNavigatedTo` method into the VM constructor. (This seems a better solution as it does away with UI first binding to an uninitialized VM. However this breaks a popular convention of initializing ViewModels once navigation is complete.)
But, making the Binding OneWay works too. (This is a simpler fix but could have other consequences.)

Further input welcome.

**Comment by allenlooplee** [comment]
@mrlancy

Thanks for the explanation. BTW, does WinTS officially support MVVMToolkit? I didn't find it in the "design pattern" options.

**Comment by mrlacey** [comment]
> @mrlancy
> 
> Thanks for the explanation. BTW, does WinTS officially support MVVMToolkit? I didn't find it in the "design pattern" options.

Support for MVVM Toolkit is ready to go. (You can already see it in the nightly builds if you want to check it out.)  
We're just waiting for it to come out of preview as WinTS doesn't include preview packages due to the risk of breaking changes causing problems in the future.

**Comment by mrlacey** [comment]
Having now reviewed all the other ways that SettingsViewModels are set up and used in all the other WinTS options, I'm still unsure what the best option is for this issue.

If this is to be addressed, then the change needs to only modify what is generated for the Settings page when using MVVMLight. 
The current templates do not include anything (regular or composition templates) for this specific scenario. If they did already exist I'd look for a solution that could be added there.

The simplest solution I can see is to add a new composition template for WPF|MVVMLight|Page.Settings that does a replace on SettingsPage.xaml and replaces `IsChecked="{Binding Theme, ` with `IsChecked="{Binding Theme, Mode=OneWay, `.
If replace doesn't work with partial lines (I can't remember if it needs full lines or not) then there will need to be 3 entries for the file (one for each theme).
This doesn't feel like a perfect solution but will be good enough given the deprecation of support for the framework and the low number of people this impacts.
I'm keen to hear other opinions on this before anything is implemented.

**Comment by sibille** [comment]
Synced with @mvegaca offline: the reason he was using the command to update the theme was that you can't execute async code from a Setter (there is no async code currently, but this pattern allows you to include async code). 
We think we should use Mode=OneWay for all frameworks here as the command is used to update values from View to ViewModel and we only need to sync from ViewModel to View.
This seems to be a permutation of https://github.com/microsoft/WindowsTemplateStudio/issues/1053

**Comment by JeanRoca** [comment]
Team has decided we will be including a fix for this issue in the 3.10 release.
